humility-core/src/hubris.rs

```diff
@@ -3034,6 +3034,38 @@ impl HubrisArchive {
         !self.modules.is_empty()
     }
 
+    ///
+    /// Takes a list of potentially similar types and deduplicates the list.
+    ///
+    fn dedup(&self, goffs: Vec<HubrisGoff>) -> Result<Vec<HubrisGoff>> {
+        let mut cmp = Vec::new();
+        let mut out = Vec::new();
+
+        if goffs.len() <= 1 {
+            Ok(goffs)
+        } else {
+            for goff in goffs {
+                cmp.push((goff, self.lookup_type(goff)?));
+            }
+
+            for i in 0..cmp.len() {
+                let mut dup = false;
+
+                for j in i + 1..cmp.len() {
+                    if !cmp[i].1.differs(self, &cmp[j].1)? {
+                        dup = true;
+                        break;
+                    }
+                }
+
+                if !dup {
+                    out.push(cmp[i].0);
+                }
+            }
```

Comment:
We could remove (some) explicit indexing here:
```suggestion
            for (i, a) in cmp.iter().enumerate() {
                let mut dup = false;

                for b in &cmp[i + 1..] {
                    if !a.1.differs(self, &b.1)? {
                        dup = true;
                        break;
                    }
                }

                if !dup {
                    out.push(a.0);
                }
            }
```

Reply:
I don't think I've ever used it, but `Vec::dedup_by()` exists; I don't _think_ we can use it here because `differs` is fallible, though?